[inductor] Insert triton barrier before storing to inplace buffers

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100769

The linked issue demonstrates a triton bug where a load broadcasted
over multiple warps may see the result of a store that happens later
in the triton program. The workaround is to add a barrier before
storing, which enforces that all warps have already read the data.

e.g. in `test_embedding_var_mean` we now generate:
```python
    tl.debug_barrier()
    tl.store(in_out_ptr1 + (tl.broadcast_to(x0, [XBLOCK, 1])), tmp17, None)
```



cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[inductor] Insert triton barrier before storing to inplace buffers

The linked issue demonstrates a triton bug where a load broadcasted
over multiple warps may see the result of a store that happens later
in the triton program. The workaround is to add a barrier before
storing, which enforces that all warps have already read the data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Insert triton barrier before storing to inplace buffers"


The linked issue demonstrates a triton bug where a load broadcasted
over multiple warps may see the result of a store that happens later
in the triton program. The workaround is to add a barrier before
storing, which enforces that all warps have already read the data.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel reviewed (approved):
Any perf impact here?

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1167,6 +1167,19 @@ def store(self, name, index, value, mode=None):
+        is_inplace = name in self.args.inplace_buffers
+        is_broadcasted = not set.issubset(
+            set(self.range_tree_nodes.keys()), original_index.free_symbols
+        )
```
It'd be nice to have this in `utils.py`. I need exactly this for another PR.

peterbell10 replied:
@lezcano I hope you don't mind I'll just merge this as is and I'll trust you to refactor it.

lezcano replied:
for sure

peterbell10 committed:
Update on "[inductor] Insert triton barrier before storing to inplace buffers"


The linked issue demonstrates a triton bug where a load broadcasted
over multiple warps may see the result of a store that happens later
in the triton program. The workaround is to add a barrier before
storing, which enforces that all warps have already read the data.

e.g. in `test_embedding_var_mean` we now generate:
```python
    tl.debug_barrier()
    tl.store(in_out_ptr1 + (tl.broadcast_to(x0, [XBLOCK, 1])), tmp17, None)
```



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
